<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>NPM-sample-qc &mdash; NPM-sample-qc  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Functions" href="metrics.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="#" class="icon icon-home"> NPM-sample-qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Metric definitions</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="metrics.html">Functions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">NPM-sample-qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
      <li>NPM-sample-qc</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="npm-sample-qc">
<h1>NPM-sample-qc<a class="headerlink" href="#npm-sample-qc" title="Permalink to this heading"></a></h1>
<p>NPM-sample-qc is a <a class="reference external" href="https://www.nextflow.io/">Nextflow</a> workflow to obtain QC metrics from single-sample WGS results. It has been created to support QC efforts within the National Precision Medicine programme in Singapore (NPM). In collaboration with <a class="reference external" href="https://www.ga4gh.org/">GA4GH</a> , the <a class="reference external" href="https://www.ga4gh.org/product/wgs-quality-control-standards/">WGS_Quality_Control_Standards</a> was formed with this workflow as first reference implementation of the standards.</p>
<section id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><a class="reference external" href="https://www.nextflow.io/docs/latest/getstarted.html#installation">Install Nextflow</a> v23.04 or higher</p></li>
<li><p><a class="reference external" href="https://docs.docker.com/get-docker/">Install Docker</a></p></li>
<li><p>Install and configure <a class="reference external" href="https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html">AWS CLI</a></p></li>
</ul>
</section>
<section id="quick-start">
<h2>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this heading"></a></h2>
<p>Clone this repository</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">git</span><span class="nd">@github</span><span class="o">.</span><span class="n">com</span><span class="p">:</span><span class="n">c</span><span class="o">-</span><span class="n">BIG</span><span class="o">/</span><span class="n">NPM</span><span class="o">-</span><span class="n">sample</span><span class="o">-</span><span class="n">qc</span><span class="o">.</span><span class="n">git</span>
</pre></div>
</div>
<p>Run workflow on 45Mbp region around AKT1 gene, 30X, of sample NA12878 from the 1000 Genomes Phase 3 Reanalysis with DRAGEN 3.7</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Move to test folder</span>
<span class="n">cd</span> <span class="n">tests</span><span class="o">/</span><span class="n">NA12878</span><span class="o">-</span><span class="n">chr14</span><span class="o">-</span><span class="n">AKT1_1000genomes</span><span class="o">-</span><span class="n">dragen</span><span class="o">-</span><span class="mf">3.7.6</span><span class="o">/</span>
<span class="c1"># Run the workflow</span>
<span class="n">sh</span> <span class="n">run</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<p>This creates <cite>output</cite> directory with the results that can be compared to the content of <cite>output_certified</cite></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">diff</span> <span class="n">output_certified</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">metrics</span><span class="o">/</span><span class="n">NA12878</span><span class="o">-</span><span class="n">chr14</span><span class="o">-</span><span class="n">AKT1</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">json</span> <span class="n">output</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">metrics</span><span class="o">/</span><span class="n">NA12878</span><span class="o">-</span><span class="n">chr14</span><span class="o">-</span><span class="n">AKT1</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">json</span>
</pre></div>
</div>
<p>Please refer to the workflow help for more information on its usage and access to additional options:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">NPM</span><span class="o">-</span><span class="n">sample</span><span class="o">-</span><span class="n">qc</span><span class="o">/</span><span class="n">main</span><span class="o">.</span><span class="n">nf</span> <span class="o">--</span><span class="n">help</span>
</pre></div>
</div>
</section>
<section id="understanding-the-workflow">
<h2>Understanding the workflow<a class="headerlink" href="#understanding-the-workflow" title="Permalink to this heading"></a></h2>
<section id="resources">
<h3>Resources<a class="headerlink" href="#resources" title="Permalink to this heading"></a></h3>
<p>The workflow requires the following resources given in the <code class="docutils literal notranslate"><span class="pre">conf/resources.config</span></code></p>
<ul class="simple">
<li><p><em>N-regions reference file</em>, used as an input for computing “non-gap regions autosome” coverages (mosdepth).</p>
<ul>
<li><p>Gaps in the GRCh38 (hg38) genome assembly, defined in the AGP file delivered with the sequence, are being closed during the finishing process on the human genome. GRCh38 (hg38) genome assembly still contains the following principal types of gaps:</p>
<ul>
<li><p>short_arm - short arm gaps (count: 5; size range: 5,000,000 - 16,990,000 bases)</p></li>
<li><p>heterochromatin - heterochromatin gaps (count: 11; size range: 20,000 - 30,000,000 bases)</p></li>
<li><p>telomere - telomere gaps (count: 48; all of size 10,000 bases)</p></li>
<li><p>contig - gaps between contigs in scaffolds (count: 285; size range: 100 - 400,000 bases)</p></li>
<li><p>scaffold - gaps between scaffolds in chromosome assemblies (count: 470; size range: 10 - 624,000 bases)</p></li>
</ul>
</li>
<li><p>Gaps in the GRCh38 (hg38) genome assembly were downloaded from <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/gap.txt.gz">http://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/gap.txt.gz</a> (<code class="docutils literal notranslate"><span class="pre">2019-03-11</span> <span class="pre">09:51,</span> <span class="pre">12K</span></code>).</p></li>
</ul>
</li>
<li><p><em>Human Reference Genome FASTA file</em>, used as an input for multiple tools.</p></li>
</ul>
<p>This file can be downloaded from <code class="docutils literal notranslate"><span class="pre">s3://broad-references/hg38/v0/Homo_sapiens_assembly38.fasta</span></code>.</p>
<ul class="simple">
<li><p><em>FASTA file index</em>. This file can be downloaded from <code class="docutils literal notranslate"><span class="pre">s3://1000genomes-dragen-3.7.6/references/fasta/hg38.fa.fai</span></code> and not required to be specified in the config. The workflow will look fasta index file in a folder the fasta file is present.</p></li>
</ul>
</section>
<section id="inputs">
<h3>Inputs<a class="headerlink" href="#inputs" title="Permalink to this heading"></a></h3>
<p>Input requirements can be split into two categories:</p>
<ul class="simple">
<li><p><strong>Generic workflow settings</strong> specify parameters that will not vary from run to run, e.g.</p></li>
</ul>
<p>Nextflow profile declarations, trace/timeline/report/dag options, output structure and paths to data resources.
See <code class="docutils literal notranslate"><span class="pre">nextflow.config</span></code> for additional details.</p>
<ul class="simple">
<li><p><strong>Sample-specific settings</strong> contain paths to WGS data for a given sample, namely BAM/CRAM.</p></li>
</ul>
<p>The workflow expects the BAM/CRAM index (bai/crai) to be present in the same location.
See <code class="docutils literal notranslate"><span class="pre">tests/NA12878-chr14-AKT1_1000genomes-dragen-3.7.6/params.yml</span></code> for an example.</p>
<p>If accessing AWS S3 public resources (for example reference genome .fa, .fai or bam/cram) without any AWS user credential,
append <code class="docutils literal notranslate"><span class="pre">aws_no_sign_request:</span> <span class="pre">true</span></code> to your parameter list (<code class="docutils literal notranslate"><span class="pre">params.yaml</span></code> or commandline argument).
See <code class="docutils literal notranslate"><span class="pre">tests/NA12878_1000genomes-dragen-3.7.6/params.yml</span></code> for an example.</p>
</section>
<section id="outputs">
<h3>Outputs<a class="headerlink" href="#outputs" title="Permalink to this heading"></a></h3>
<p>Upon completion, the workflow will create the following files in the <code class="docutils literal notranslate"><span class="pre">outdir</span></code> directory:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">outdir</span><span class="o">/</span>
    <span class="n">pipeline_info</span><span class="o">/</span>    <span class="c1"># dag, timeline and trace files</span>
        <span class="n">dag</span><span class="o">.</span><span class="n">pdf</span>
        <span class="n">report</span><span class="o">.</span><span class="n">html</span>
        <span class="n">timeline</span><span class="o">.</span><span class="n">html</span>
        <span class="n">trace</span><span class="o">.</span><span class="n">txt</span>
    <span class="n">results</span><span class="o">/</span>          <span class="c1"># final metrics.json and intermediate outputs</span>
        <span class="n">metrics</span><span class="o">/&lt;</span><span class="n">sample_id</span><span class="o">&gt;.</span><span class="n">metrics</span><span class="o">.</span><span class="n">json</span>
        <span class="n">mosdepth</span><span class="o">/</span>
        <span class="n">multiqc</span><span class="o">/</span>
        <span class="n">picard_collect_multiple_metrics</span><span class="o">/</span>
        <span class="n">samtools</span><span class="o">/</span>
</pre></div>
</div>
<p>If <code class="docutils literal notranslate"><span class="pre">keep_workdir</span></code> has been specified, the contents of the Nextflow work directory (<code class="docutils literal notranslate"><span class="pre">work-dir</span></code>) will also be preserved.</p>
</section>
<section id="docker-image">
<h3>Docker image<a class="headerlink" href="#docker-image" title="Permalink to this heading"></a></h3>
<p>By default the workflow pull the docker image from dockerhub. However you can also build docker image locally</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Move to containers</span>
<span class="n">cd</span> <span class="n">NPM</span><span class="o">-</span><span class="n">sample</span><span class="o">-</span><span class="n">qc</span><span class="o">/</span><span class="n">containers</span>
<span class="c1"># Build docker image locally</span>
<span class="n">sh</span> <span class="n">build_npm</span><span class="o">-</span><span class="n">sample</span><span class="o">-</span><span class="n">qc_docker_image</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
</section>
</section>
<section id="workflow-logic">
<h2>Workflow logic<a class="headerlink" href="#workflow-logic" title="Permalink to this heading"></a></h2>
<p>We provide a schematic representation of the workflow in the figure below:</p>
<img src="./npm-sample-qc-overview.PNG" width="500px"/><p>In a nutshell, this workflow generates QC metrics from single-sample WGS results in three stages: <strong>metrics calculation</strong>, <strong>parsing of intermediate outputs</strong> and <strong>generation of a final report</strong>. This makes it possible to take full advantage of the parallelisation capabilities of Nextflow, allows users to leverage third-party tools or add custom scripts, and enables auto-documentation of metrics from code comments.</p>
<p><strong>Metrics calculation</strong></p>
<p>The current workflow combines widely-used third-party tools (samtools, picard, mosdepth) and custom scripts. Full details on which processes are run/when can be found in the actual workflow definition (<code class="docutils literal notranslate"><span class="pre">main.nf</span></code>). We also provide an example dag for a more visual representation (<code class="docutils literal notranslate"><span class="pre">tests/NA12878_1000genomes-dragen-3.7.6/dag.pdf</span></code>).</p>
<p><strong>Metrics parsing</strong></p>
<p>Next, output files from each individual tool are parsed and combined into a single json file. This is done by calling <code class="docutils literal notranslate"><span class="pre">bin/multiqc_plugins/multiqc_npm/</span></code>, a MultiQC plugin that extends the base tool to support additional files.</p>
<p><strong>Metrics reporting</strong></p>
<p>Finally, the contents of the MultiQC json are formatted into a final metrics report, also in json format. The reporting logic lives in the <code class="docutils literal notranslate"><span class="pre">bin/compile_metrics.py</span></code> script, and whilst its contents are simple, it enables automatic documentation of metric definitions from code comments (see the <strong>Metric definitions</strong> section).</p>
</section>
<section id="metric-definitions">
<h2>Metric definitions<a class="headerlink" href="#metric-definitions" title="Permalink to this heading"></a></h2>
<p>The full list of metrics reported by this workflow and details on how they’ve been calculated can be found <a class="reference external" href="https://c-big.github.io/NPM-sample-qc/metrics.html">here</a>.</p>
<p>When needed, page contents can be updated by running the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Install sphinx</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">sphinx_rtd_theme</span> <span class="n">sphinx_automodapi</span>
<span class="c1"># Move to doc source</span>
<span class="n">cd</span> <span class="n">docsrc</span>
<span class="c1"># Build the doc</span>
<span class="o">./</span><span class="n">build</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Metric definitions</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="metrics.html">Functions</a></li>
</ul>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="metrics.html" class="btn btn-neutral float-right" title="Functions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2023, GIS.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>