<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Functions &mdash; NPM-sample-qc  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="count_deletions" href="api/metrics.count_deletions.html" />
    <link rel="prev" title="NPM-sample-qc" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> NPM-sample-qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Metric definitions</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.count_deletions.html">count_deletions</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.count_insertions.html">count_insertions</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.count_snvs.html">count_snvs</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.cross_contamination_rate.html">cross_contamination_rate</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.insert_size_std_deviation.html">insert_size_std_deviation</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.mad_autosome_coverage.html">mad_autosome_coverage</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.mean_autosome_coverage.html">mean_autosome_coverage</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.mean_insert_size.html">mean_insert_size</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.pct_autosomes_15x.html">pct_autosomes_15x</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.pct_reads_mapped.html">pct_reads_mapped</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.pct_reads_properly_paired.html">pct_reads_properly_paired</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.ratio_heterozygous_homzygous_indel.html">ratio_heterozygous_homzygous_indel</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.ratio_heterozygous_homzygous_snv.html">ratio_heterozygous_homzygous_snv</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.ratio_insertion_deletion.html">ratio_insertion_deletion</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.ratio_transitions_transversions.html">ratio_transitions_transversions</a></li>
<li class="toctree-l2"><a class="reference internal" href="api/metrics.yield_bp_q30.html">yield_bp_q30</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">NPM-sample-qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Functions</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/metrics.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="functions">
<h1>Functions<a class="headerlink" href="#functions" title="Permalink to this headline"></a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="api/metrics.count_deletions.html#metrics.count_deletions" title="metrics.count_deletions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_deletions</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The ratio between number of insertion and deletion in VCF, only in autosomal regions, high quality variants.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/metrics.count_insertions.html#metrics.count_insertions" title="metrics.count_insertions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_insertions</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The number of variant type indels only insertions in VCF, only in autosomal regions, high quality variants.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/metrics.count_snvs.html#metrics.count_snvs" title="metrics.count_snvs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_snvs</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The number of variant type SNVs in VCF, only in autosomal regions, high quality variants.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/metrics.cross_contamination_rate.html#metrics.cross_contamination_rate" title="metrics.cross_contamination_rate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cross_contamination_rate</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: Estimation of inter-sample contamination rate of short paired-end sequencing high quality, non duplicated reads, primary alignments, mapped on GRCh38 assembly.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/metrics.insert_size_std_deviation.html#metrics.insert_size_std_deviation" title="metrics.insert_size_std_deviation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert_size_std_deviation</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The insert size standard deviation of short paired-end sequencing high quality reads, primary alignments, mapped on GRCh38 assembly.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/metrics.mad_autosome_coverage.html#metrics.mad_autosome_coverage" title="metrics.mad_autosome_coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mad_autosome_coverage</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The median absolute deviation of sequencing coverage derived from short paired-end sequencing high quality, non duplicated reads, primary alignments, achieving a base quality of 20 or greater and mapping quality of 20 or greater, in autosomes non gap regions of GRCh38 assembly.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/metrics.mean_autosome_coverage.html#metrics.mean_autosome_coverage" title="metrics.mean_autosome_coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mean_autosome_coverage</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The mean sequencing coverage derived from short paired-end sequencing high quality, non duplicated reads, primary alignments, achieving a base quality of 20 or greater and mapping quality of 20 or greater, in autosomes non gap regions of GRCh38 assembly.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/metrics.mean_insert_size.html#metrics.mean_insert_size" title="metrics.mean_insert_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mean_insert_size</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The mean insert size of short paired-end sequencing high quality reads, primary alignments, mapped on GRCh38 assembly.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/metrics.pct_autosomes_15x.html#metrics.pct_autosomes_15x" title="metrics.pct_autosomes_15x"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pct_autosomes_15x</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The percentage of bases attaining at least 15X sequencing coverage in short paired-end sequencing high quality, non duplicated reads, primary alignments, achieving a base quality of 20 or greater and mapping quality of 20 or greater, in autosomes non gap regions of GRCh38 assembly.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/metrics.pct_reads_mapped.html#metrics.pct_reads_mapped" title="metrics.pct_reads_mapped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pct_reads_mapped</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The percentage of short paired-end sequencing high quality reads, primary alignments, mapped on GRCh38 assembly.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/metrics.pct_reads_properly_paired.html#metrics.pct_reads_properly_paired" title="metrics.pct_reads_properly_paired"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pct_reads_properly_paired</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The percentage of short paired-end sequencing high quality, properly paired reads, primary alignments, mapped on GRCh38 assembly.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/metrics.ratio_heterozygous_homzygous_indel.html#metrics.ratio_heterozygous_homzygous_indel" title="metrics.ratio_heterozygous_homzygous_indel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ratio_heterozygous_homzygous_indel</span></code></a>(mqc, ...)</p></td>
<td><p>Description: The ratio of heterozygous and homozygous variant type indels in VCF, only in autosomal regions, high quality variants.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/metrics.ratio_heterozygous_homzygous_snv.html#metrics.ratio_heterozygous_homzygous_snv" title="metrics.ratio_heterozygous_homzygous_snv"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ratio_heterozygous_homzygous_snv</span></code></a>(mqc, ...)</p></td>
<td><p>Description: The ratio of heterozygous and homozygous variant type SNVs in VCF, only in autosomal regions, high quality variants.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/metrics.ratio_insertion_deletion.html#metrics.ratio_insertion_deletion" title="metrics.ratio_insertion_deletion"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ratio_insertion_deletion</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The ratio between number of insertion and deletion in VCF, only in autosomal regions, high quality variants.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/metrics.ratio_transitions_transversions.html#metrics.ratio_transitions_transversions" title="metrics.ratio_transitions_transversions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ratio_transitions_transversions</span></code></a>(mqc, ...)</p></td>
<td><p>Description: The ratio of transitions and transversions of bi-allelic SNVs in VCF, only in autosomal regions, high quality variants.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/metrics.yield_bp_q30.html#metrics.yield_bp_q30" title="metrics.yield_bp_q30"><code class="xref py py-obj docutils literal notranslate"><span class="pre">yield_bp_q30</span></code></a>(mqc, biosample_id)</p></td>
<td><p>Description: The number of bases in short paired-end sequencing high quality reads, primary alignments, achieving a base quality score of 30 or greater (Phred scale).</p></td>
</tr>
</tbody>
</table>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="NPM-sample-qc" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api/metrics.count_deletions.html" class="btn btn-neutral float-right" title="count_deletions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2023, GIS.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>